src/WinRT.Runtime/Interop/IContextCallback.cs

```diff
@@ -67,15 +62,42 @@ static int InvokeCallback(ComCallData* comCallData)
                 }
             }
 
-            Guid iid = IID.IID_ICallbackWithNoReentrancyToApplicationSTA;
+            ComCallData comCallData;
+            comCallData.dwDispid = 0;
+            comCallData.dwReserved = 0;
+
+            CallbackData.PerThreadObject = state;
+
+            int hresult;
+
+            // We use a thread local static field to efficiently store the state that's used by the callback. Note that this
+            // is safe with respect to reentrancy, as the target callback will never try to switch back on the original thread.
+            // We're only ever switching once on the original context, only to release the object reference that is passed as

```

Comment:
Release can run arbitrary code, is that right? I do not see what guarantees that the arbitrary code cannot switch threads at will.

Also, this seems to be used for more than just releasing the object reference: https://github.com/microsoft/CsWinRT/blob/93482af78df5c153ddc709b027c342b90a83ea86/src/WinRT.Runtime/ObjectReference.cs#L751-L760

It would be more correct by construction to implement this as usual pool - clear the cached instance when it is rented, so that there is no way for one instance to be rented multiple times at the same time.

Reply:
I would have assumed that `IContextCallback::ContextCallback` with `ICallbackWithNoReentrancyToApplicationSTA` means it shouldn't be able to run any arbitrary code on this thread as a result of that call, and nothing else from when we set the field to when we clear it seems problematic. If that thing doesn't guarantee this however, then something should be done to allow multiple current things at once potentially existing (e.g., using a `List<object>` or similar would do the job).

Reply:
Or do you mean the callback might call back into this thread & somehow end up back here? That would make sense I suppose. Is that possible @Sergio0694?

Reply:
> "as long as the callback is called before anything else silly happens, since it immediately reads the field

That's actually a good point. Even if we somehow recursively ended up here (which I don't believe is possible), each callback would've already read the target state before invoking the user-provided callback anyway.

> "Release can run arbitrary code"

Not really, I mean yes from the point of view of this API, but the context callback is only ever used internally to release `IObjectReference` objects, which are implemented in CsWinRT only. And we're only ever using these to pass the `Release` pointers which simply do a release on the tracker ref and the native object on the target context:

https://github.com/microsoft/CsWinRT/blob/93482af78df5c153ddc709b027c342b90a83ea86/src/WinRT.Runtime/ObjectReference.cs#L955-L967